### DISCUSSION TITLE
Partial accept suggestion (in VSCode)

### DISCUSSION THREAD

**Discussion by roman-balzer** [community/community#7331] [Copilot Conversations] [answered]
Sometimes Copilot gives suggestions, which are close but not quite right. Would be awesome if I could accept it until the 'not quite right' part begins.

What I type + Suggestion
let d _efaultDeviceSelected = '... selected'_

What I want to write
let defaultDeviceInvalid = '... invalid'

Lets say I have a shortcut MODKEY + Arrow Right an accept until the diff per character
let defaultDevice _Selected = '... selected'_    // now type I
let defaultDeviceI _nvalid = '... invalid'_    // now TAB and im happy

Also this could be used with CTRL + MODKEY + Arrow Right to partially accept whole words.

**Comment by marlonrichert** [comment]
I would love for Copilot to let me partially accept suggestions simply by moving the cursor, similar to how this works in [`fish` autosuggestions](https://fishshell.com/docs/current/interactive.html?highlight=autosuggestions#autosuggestions) and the [`zsh-autosuggestions` plugin](https://github.com/zsh-users/zsh-autosuggestions#usage).

* `fish`
  > To accept the first suggested word, press <kbd>Alt</kbd>+<kbd>→</kbd> or <kbd>Alt</kbd>+<kbd>F</kbd>.
* `zsh-autosuggestions`
  > If you invoke the `forward-word` widget, it will partially accept the suggestion up to the point that the cursor moves to.

**Reply by roman-balzer** [reply]
I don't think using just <kbd>→</kbd> or <kbd>End</kbd> without a mod key is a good idea. Copilot sometimes also suggest in the middle of a line, instead of at the end of a link. So if I just move around the code with <kbd>→</kbd> or <kbd>End</kbd> and Copilot happens to be fast enough to suggest something, I could accidently autocomplete something, I don't want. That already happens sometimes for me with <kbd>Tab</kbd>

**Reply by marlonrichert** [reply]
Sorry, that wasn't really what I was after. I've edited my post to make it more clear what I'm actually suggesting. Please have another look.

**Comment by jheindo1** [comment]
I think partial accept would be cool, I don't have a good suggestion for shortcuts or how exactly it should work, but I often find myself accepting the solution and immediately deleting some parts.
At least for multiline suggestions it could be useful to accept them line by line, so I can accept 3 out of 5 lines for example in a more convenient way.

**Reply by scpedicini** [reply]
This would be especially useful, often times I find that I only really want the completion of the current line, and not the expansive multi-line snippet. We need a hotkey that lets you quickly take only the first line of a suggestion and discards the rest of the block.

**Reply by unlocomqx** [reply]
> This would be especially useful, often times I find that I only really want the completion of the current line, and not the expansive multi-line snippet. We need a hotkey that lets you quickly take only the first line of a suggestion and discards the rest of the block.

I think it's better to advance word by word (check my vid) because sometimes I'd like to pick only the first half of the line and not even the whole line

**Reply by scpedicini** [reply]
@unlocomqx I can't think of a reason not to have both as *optionable* mappable keys:
- Alt-Tab (Take word by word)
- Cmd-Tab (Take entire line)

**Reply by unlocomqx** [reply]
Yes, it's a good idea to have both possibilities

**Reply by hXtreme** [reply]
I think both Alt-Tab and Cmd-Tab would conflict.

**Reply by unlocomqx** [reply]
> I think both Alt-Tab and Cmd-Tab would conflict.

Alt -> (Alt + Right arrow) would be better for word by word
And maybe Alt + Shift + Right arrow for line by line

**Reply by scpedicini** [reply]
> I think both Alt-Tab and Cmd-Tab would conflict.

I mean this isn't really relevant because in any IDE they would be mappable so we don't really need to worry about the exact shortcuts.

**Reply by scpedicini** [reply]
If anything I would prefer that the co-pilot team focus their efforts on adding functionality (like word by word, or line by line) but leave the hot keys unmapped by default.

Everybody is going to have their own personal preferences for what keys they want to use (Windows, Mac, Linux) and you're never going to be able to satisfy everyone.

**Comment by alvitawa** [comment]
I think this is a great idea, especially useful when writing normal text in for example markdown files, because suggestions are really long usually and wrong, but the first word is usually correct.

Very simple shortcut combination: Change Tab to do a 'short' complete (only untill the next space for example), and spam Tab or press and hold to do a 'full' completion (whole suggestion).

~~Alternatively, have a global setting that determines suggestion length (measured in words or characters) and increase/decrease with keyboard shortcuts like `Ctrl+[` and `Ctrl+]`.~~

**Reply by hXtreme** [reply]
Same! I like to use co-pilot for writing doc strings and this really hurts its usability.

**Comment by frederickjjoubert** [comment]
To add to this post, I am working with a Unity3D project and have the same problem very frequently with my C# code. It would be great to be able to `tab` through up to a certain point instead of accepting the whole suggestion and then having to go back and change things, as that ends up slowing me down far more using github copilot than without it.

**Reply by SaladinoGW** [reply]
Me too

**Reply by hXtreme** [reply]
same! plus its frustrating knowing that you're doing something manually when the co-pilot's suggestion is kinda correct but unusable.

**Comment by RichardMM** [comment]
Really looking foward to this feature. Currently I opt out of the entire line suggested only to retype it identically for several words

**Comment by unlocomqx** [comment]
It would be kinda similar to zsh-autosuggestions

https://user-images.githubusercontent.com/793712/177937730-c75e8050-83ff-4d1b-b178-af1b53d0fb68.mp4

**Comment by danizil** [comment]
Really important feature especially for a language model

**Comment by hXtreme** [comment]
I'd really want the ability to accept word-by-word.

I like to use copilot when writing docs, the first line or so of suggestion is basically on point but then after that its irrelevant stuff.
If I accept the whole thing then very often I end up with not just incorrect docs but random code sneaks in.
I have to go back and manually edit a bunch of lines, or alternatively manually write a bunch of lines that co-pilot suggested.

More so than tedium, its actually frustrating, knowing that what I want to write is right there but I can't autocomplete it.

**Comment by clementpellen** [comment]
+1 
Word-by-word accepting could be very usefull

**Comment by Vrajs16** [comment]
Any update on this? This would really help in coding to autocomplete, however the total final solution is not correct all the time.

**Comment by carlwr** [comment]
I really miss this as well - feels like I need it all the time. I also note that there are several Stack Overflow questions asked about such a feature.

Hoping for this to be implemented . In the meantime, would anyone have a suggestion on a workaround (in my case, for VS Code), that wouldn't require this functionality natively in Copilot?

**Comment by cmp-nct** [comment]
It was unbearable for me, 95% of the code suggested was garbage.
I'm afraid this won't work with your IDE but for IntelliJ users my fix works: https://github.com/cmp-nct/Stewardess
Maybe someone ports it or creates a similar one.

**Reply by cpriest** [reply]
I very frequently have to press escape to cancel co-pilot and have to remember to press enter to accept the ide autocomplete.

**Comment by cpriest** [comment]
This could also be handled the way IntelliJ does template shortcuts and other "insertions."

https://user-images.githubusercontent.com/159323/200864869-28282320-5fec-4c57-95b7-5b9cd513b3d8.mp4

Pressing the tab key selects the new block to be accepted or edited.

**Comment by growthboot** [comment]
This would make a big difference to my efficiency!

**Comment by BoD** [comment]
See also #12426.

**Comment by NicoWeio** [comment]
@BoD linked to it, but let's write it out here:

🎉 You should be able to use <kbd>Ctrl</kbd>+<kbd>→</kbd> on versions ≥1.74.0. 🎉

Here's the PR: https://github.com/microsoft/vscode/pull/166956

**Reply by jianghaoBUAA** [reply]
LGTM

**Comment by BoD** [comment]
In the IntelliJ (and IDEs based on it, like Android Studio, etc.) version of the plugin, this is now also available 🎉

<img width="342" alt="Screenshot 2023-11-22 at 16 57 34" src="https://github.com/community/community/assets/372852/6626f5da-3afb-43e6-b861-68f32e7c450e">

**Comment by corneliusroemer** [comment, marked as answer]
Accepting line by line, as well as word by word is now possible.

Word-by-word is enabled by default with keybinding `cmd/ctrl+right`, line-by-line does not have a keybinding by default, but this can be added.

To enable line-by-line acceptance of suggestions, you need to create a custom keybinding, e.g. `ctrl+shift+right_arrow`, and map it to the `editor.action.inlineSuggest.acceptNextLine` command.

Paste this into the `keybindings.json` file (which you can find through the omnibox):

```json
{
  "key": "shift+ctrl+right"    
  "command": "editor.action.inlineSuggest.acceptNextLine",
  "when": "inlineSuggestionVisible && !editorReadonly"
}
```

In this demo I use different keybindings `shift+cmd+l` for the same action:

[![enter image description here][1]][1]

  [1]: https://i.stack.imgur.com/2RhTy.gif

**Reply by roman-balzer** [reply]
Just recently saw the shortcuts, thanks for the reminder. I accepted your reply as answer

**Reply by flying318** [reply]
Just saw the soluiton and thanks for this config. I successfully applied it in my vscode and it worked. 😄
